README
Please refer to the online documentation at https://mdpdocs.readthedocs.io
The legacy documentation is still available at http://mdp-toolkit.sourceforge.net
The source code is available at https://github.com/mdp-toolkit/mdp-toolkit


CHANGES
-------------------------------------------------------------------------------
MDP-3.6.1:
2018-05-29: DOC: Point to the new website at https://mdp-toolkit.github.io.

-------------------------------------------------------------------------------
MDP-3.6:
2020-04-18: NEW: support libsvm for Python 3 libsvm has been updated to be
            compatible with Python >= 3.4. Python 2 support has been dropped,
            so we don't test it anymore.
2020-04-18: ERF: restore Python 2 compatibility and add a matrix entry for it
            in travis
2019-18-12: FIX: SciPy 1.4 support.
2019-13-12: NEW: VartimeSFANode supports SFA on data with non-constant time
            increments. This yields an SFA implementation that is robust
            against missing data.
2019-01-20: NEW: Added a new family of expansion nodes, including Legendre,
            Hermite and Chebyshev polynomials. A facility for easily including
            recursively defined 1-dimensional expansion functions powers these
            nodes, applying them to N-dimensional data by tensor basis
            construction. The new expansion nodes are numerically more stable
            than ordinary monomials and permit higher expansion degrees.
            Examples where numerically stable expansion is beneficial:
            https://github.com/NiMlr/pyfunctionbases
            https://arxiv.org/abs/1805.08565, section 4.2.1, Figure 11 and A.2.
2019-01-19: FIX: pytest Python 3.7 compatibility
2018-07-18: FIX: PEP 479 compatibility
2018-06-15: NEW: Added SFA-based supervised learning, specifically graph based
            SFA nodes GSFANode and iGSFANode and hierarchical GSFA (HGSFA).
            For further information, see:
            Escalante-B A.-N., Wiskott L, "How to solve classification and
            regression problems on high-dimensional data with a supervised
            extension of Slow Feature Analysis". Journal of Machine Learning
            Research 14:3683-3719, 2013.
            http://www.jmlr.org/papers/volume14/escalante13a/escalante13a.pdf
            https://github.com/AlbertoEsc/cuicuilco
2018-05-29: DOC: Revamped the documentation for Sphinx-autodoc compatibility.
            MDP documentation is now available on readthedocs:
            https://mdpdocs.readthedocs.io/en/latest
2018-01-18: NEW: Added solvers to SFA-node that are robust against rank deficit
            in the covariance matrix. When data contained linear redundancies,
            like easily introduced by e.g. insufficient cropping of black bars
            of a video, this usually resulted in
            SymeigException ('Covariance matrices may be singular').
            Now it provides simple instructions for robust processing of such
            data by setting a new flag in SFANode:
            rank_deficit_method can be 'none' (default), 'reg', 'pca', 'svd',
            leveraging robust solving via regularization, PCA, SVD or LDL
            decomposition respectively. (LDL requires SciPy >= 1.0)
2017-03-11: DOC: changed reference to new mailing list
            old: mdp-toolkit-users@lists.sourceforge.net
            new: mdp-toolkit@python.org
2017-03-06: NEW: Added online mode to enable use of MDP in reinforcement
            learning. This notably includes incremental SFA as described in
            "Varun Raj Kompella, Matthew Luciw, and Juergen Schmidhuber (2012).
            Incremental Slow Feature Analysis: Adaptive Low-Complexity Slow
            Feature Updating from High-Dimensional Input Streams.
            Neural Computation."
            https://www.mitpressjournals.org/doi/abs/10.1162/NECO_a_00344
            New nodes and flows:
            OnlineNode, OnlineFlow, CircularFlow, OnlineCenteringNode,
            OnlineTimeDiffNode, OnlineFlowNode, CircularOnlineFlowNode,
            OnlineLayer, CloneOnlineLayer, SameInputOnlineLayer, MCANode,
            CCIPCANode, IncSFANode

-------------------------------------------------------------------------------
MDP-3.5:
2016-03-08: fix use of str in biflow indexing
2016-03-08: remove obsolete scripts
2016-03-07: Add wrapper function for exec Old Python versions don't allow exec
            statement in nested functions: Traceback (most recent call last):
             File "<string>", line 1, in <module>
             File "mdp/__init__.py", line 131, in <module>
               from . import utils
             File "mdp/utils/__init__.py", line 29, in <module>
               from .slideshow import (basic_css, slideshow_css, HTMLSlideShow,
             File "mdp/utils/slideshow.py", line 23, in <module>
               from . import templet
             File "mdp/utils/templet.py", line 196
               exec(code, globals, kw) SyntaxError: unqualified exec is not
            allowed in function 'expand' it is a nested function

-------------------------------------------------------------------------------
MDP-3.4:
2016-03-02: Make use of setuptools instead of distutils While we resisted as
            long as possible, it is now time to move to use setuptools
            officially. This makes MDP installable properly with pip. This
            means that our dependencies are now explicitly declared
            (numpy and future). Optional dependencies are also declared and can
            be installed using the "pip -e MDP[optionaldep]" command, where
            optionaldep can be scipy, or scikit-learn or ... Moreover,
            automatic testing using "python setup.py test" now works, which
            makes MDP finally conform this non-written standard and makes it 
            easy to run tests for clueless people.
2016-03-02: add new modules in sklearn wrappers
2016-03-02: Remove self-contained py.test script... The use of the
            self-contained script for py.test has been deprecated for a while
            and has stop working. We now require py.test to be installed to run
            the tests.

testall.py
from __future__ import print_function
# calls to os.system should be changed to subprocess.Popen!
# we don't need any temporary files and such. it is important
# to set the environment properly
# I call it like this:
# $ cd /home/tiziano/git/MDP/mdp-toolkit
# $ python testall.py /home/tiziano/python/x86_64/lib/pythonVERSION/site-packages

PARMS = {
         '3.4': ('numpy', None, 'joblib', 'scikit-learn'),
         '3.5': ('numpy', None, 'joblib', 'scikit-learn'),
         '2.7': ('scipy', None, 'parallel_python', 'libsvm', 'joblib', 'scikit-learn'),
         }

import os
import sys
import subprocess

# get from sys.argv a directory to add to pythonpath
# /path/to/pythonVERSION/dir
if len(sys.argv) > 1:
    dirbase = sys.argv[1]
else:
    dirbase = '/dev/null'

# check that we are in our git repo
conds = (os.path.exists('.git'),
         os.path.basename(os.getcwd()) == 'mdp-toolkit',
         os.path.exists('mdp'),
         os.path.exists('bimdp'),
         )
if not all(conds):
    sys.stderr.write('Not in mdp git clone!')
    sys.exit(-1)

startwd = os.getcwd()
config = '-c "import mdp; import sys; sys.stdout.write(mdp.config.info())"'

# create command line
for vers in PARMS:
    print('Running: '+vers)
    path = dirbase.replace('VERSION', vers)
    os.chdir(startwd)
    wd = os.getcwd()
    env = {'MDPNUMX': PARMS[vers][0]}
    for dep in PARMS[vers][1:]:
        print('NoDep: '+str(dep))
        if dep is not None:
            key = 'MDP_DISABLE_'+dep.upper()
        else:
            key = 'MDP_DISABLE_NONE'
        env[key] = '1'
        for pack in ('mdp', 'bimdp'): 
            cmdline_base = ('MDPNUMX='+env['MDPNUMX'],
                            key+'=1',
                            'PYTHONPATH='+path+':'+wd,
                            ' /usr/bin/python'+vers,
                            '-m', 'pytest'
                            )

            cmdline_config = (config,)
            cmdline_tests = (
                             '--capture', 'fd',
                             '-x',
                             pack,
                             ' >',
                             '/tmp/mdp_current_test',
                             '2>&1',
                             )
            # show config
            #os.system(' '.join(cmdline_base+cmdline_config))
            sys.stdout.write('\n')
            # write out command line
            #print '  '+' '.join(cmdline_base+cmdline_tests)
            exit_status = os.system(' '.join(cmdline_base+cmdline_tests))
            if exit_status != 0:
                sys.stderr.write('='*30+' FAILURE '+'='*30)
                sys.stderr.write('\nLog is in /tmp/mdp_current_test.\n')
                sys.exit(-1)
